src/MAVN.Service.CurrencyConverter.DomainServices/ConverterService.cs

```diff
@@ -1,47 +1,92 @@
-using System;
+﻿using System;
 using System.Threading.Tasks;
+using Common;
 using MAVN.Numerics;
-using MAVN.Service.CurrencyConverter.Domain.Exceptions;
+using MAVN.Service.CurrencyConverter.Domain.Models;
 using MAVN.Service.CurrencyConverter.Domain.Services;
+using Microsoft.Extensions.Caching.Distributed;
 
 namespace MAVN.Service.CurrencyConverter.DomainServices
 {
     public class ConverterService : IConverterService
     {
-        private readonly ICurrencyRateService _currencyRateService;
+        private readonly IExchangeRatesApi _exchangeRatesApi;
+        private readonly IRatesApi _ratesApi;
+        private readonly IDistributedCache _distributedCache;
         private readonly IGlobalCurrencyRateService _globalCurrencyRateService;
 
         public ConverterService(
-            ICurrencyRateService currencyRateService,
+            IExchangeRatesApi exchangeRatesApi,
+            IRatesApi ratesApi,
+            IDistributedCache distributedCache,
             IGlobalCurrencyRateService globalCurrencyRateService)
         {
-            _currencyRateService = currencyRateService;
+            _exchangeRatesApi = exchangeRatesApi;
+            _ratesApi = ratesApi;
+            _distributedCache = distributedCache;
             _globalCurrencyRateService = globalCurrencyRateService;
         }
 
         public async Task<decimal> ConvertAsync(string fromAsset, string toAsset, decimal amount)
         {
             if (fromAsset == toAsset)
-            {
                 return amount;
-            }
 
-            var currencyRate = await _currencyRateService.GetByIdAsync(fromAsset, toAsset);
+            var rate = await GetExchangeRate(fromAsset, toAsset);
 
-            if (currencyRate == null)
-                throw new EntityNotFoundException();
-
-            return amount * currencyRate.Rate;
+            return amount / rate;
         }
 
         public async Task<decimal> ConvertTokensToBaseCurrencyAsync(Money18 amount)
         {
             var globalCurrencyRate = await _globalCurrencyRateService.GetAsync();
-            
+
             if (globalCurrencyRate.Rate.Equals(0))
                 throw new InvalidOperationException("Base currency rate is 0, can't convert to tokens.");
 
-            return (decimal) (amount * (1 / globalCurrencyRate.Rate));
+            return (decimal)(amount * (1 / globalCurrencyRate.Rate));
+        }
+
+        private async Task<decimal> GetExchangeRate(string fromAsset, string toAsset)
+        {
+            var cached = await GetCachedValue(toAsset);
+
+            if (cached != null)
+            {
+                var value = cached.DeserializeJson<ExchangeRatesModel>();
+                return (decimal)value.Rates[fromAsset];
```

Comment:
Need to throw new EntityNotFoundException(); when target Asset is not found in rates collection